technical-docs/2021-05-remote-restore-redesign.md

```diff
@@ -0,0 +1,122 @@
+
+CassandraBackup objects should receive a finalizer from the backup controller. This will allow deletions of CassandraBackup objects to place a deletion timestamp, and upon reconciliation the backup controller will perform a cleanup of the corresponding backup in the storage bucket. Only after that cleanup task is completed can the CassandraBackup object see its finalizer removed and be garbage collected.
+
+The sync operation would:
+
+- List the backups in the storage bucket, through a gRPC call
+- Compare that list to the existing CassandraBackup resources
+- Create a CassandraBackup object for each missing backup
+- Delete local resources that don’t exist in the list of backups
+- Delete remote backups through gRPC for CassandraBackup objects marked for deletion, and upon success delete the resource.
+
+## CassandraBackup CRD changes
+CassandraBackup resources shouldn’t store the CassandraDatacenter spec anymore but rather the topology object generated by Medusa.  
+This object will contain the list of live nodes at the time of the backup along with their token assignments.
+Sample topology object:
+```
+{"ip-179-31-16-38.us-west-2.compute.internal": {"tokens": [-1051417974279770378, -112095142485141927, 9217379683620055011, 939597040143840721], "is_up": true}, 
+ "ip-179-31-23-54.us-west-2.compute.internal": {"tokens": [-1080182033668782299, -1156652990105327691, 9134535641516355167, 9216775765860996507], "is_up": true}, 
+ "ip-179-31-30-139.us-west-2.compute.internal": {"tokens": [-1018393144648419592, -1342532530495729638, 8846603596658721956, 8849342512812637165], "is_up": true}}
+```
+
+The CRD should also contain information that should be overridable such as the prefix, which is controlled by the namespace name. 
+
+## Medusa changes
+The tokenmap generated by Medusa will need to be augmented to provide the placement of nodes in the cluster: datacenter and rack.
+Currently Medusa is not dc/rack aware and it can probably be fairly surprising to users. It was built to backup a single DC per cluster and didn’t account for racks either.  
+Restoring from a DC/rack aware cluster to another one imposes some constraints on the source/target nodes mapping, which are currently impossible to deal with safely in a remote restore: 
+
+When using NetworkTopologyStrategy, the placement of replicas is dependent on the DC and rack layout.
+In this context, the replica for a token range will be placed on the node that owns the next token range in the next rack.
+We can't restore any node to any other node as we may end up putting several replicas in the same rack, and the nodes we'd restore to would possibly end up not containing data they're supposed be replica for.
+We can be rigid about the mapping and expect the same dc and rack names exclusively, or do a 1:1 mapping between DCs/Racks despite having different names (but expect the same number of both).
```

Comment:
I'm clearly a newbie in Medusa but dealing with heterogenous topologies sounds to me as a recipe for trouble. Shouldn't we put some constraints on what the origin and target clusters should look like? E.g. same number of nodes, 1:1 dc-rack pairing, same topology strategy... Another idea: require the user to provide the full origin-to-target mapping somehow?

Reply:
Supporting restore in cluster that have a different topology is non trivial indeed but that's really what you expect from a backup/restore solution.

As you found out in your next comment, we'll limit the supported cases here as I'm not crazy about using the sstableloader...